pkg/microservice/aslan/core/workflow/service/webhook/codehub.go

```diff
@@ -0,0 +1,176 @@
+package webhook
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"go.uber.org/zap"
+
+	commonmodels "github.com/koderover/zadig/pkg/microservice/aslan/core/common/repository/models"
+	commonrepo "github.com/koderover/zadig/pkg/microservice/aslan/core/common/repository/mongodb"
+	gitservice "github.com/koderover/zadig/pkg/microservice/aslan/core/common/service/git"
+	"github.com/koderover/zadig/pkg/setting"
+	"github.com/koderover/zadig/pkg/tool/codehub"
+	e "github.com/koderover/zadig/pkg/tool/errors"
+)
+
+func ProcessCodehubHook(payload []byte, req *http.Request, requestID string, log *zap.SugaredLogger) error {
+	token := req.Header.Get("X-Codehub-Token")
+	secret := gitservice.GetHookSecret()
+
+	if secret != "" && token != secret {
+		return errors.New("token is illegal")
+	}
+
+	eventType := codehub.HookEventType(req)
+	event, err := codehub.ParseHook(eventType, payload)
+	if err != nil {
+		return err
+	}
+
+	forwardedProto := req.Header.Get("X-Forwarded-Proto")
+	forwardedHost := req.Header.Get("X-Forwarded-Host")
+	baseURI := fmt.Sprintf("%s://%s", forwardedProto, forwardedHost)
+
+	var pushEvent *codehub.PushEvent
+	var mergeEvent *codehub.MergeEvent
+	var errorList = &multierror.Error{}
+
+	switch event := event.(type) {
+	case *codehub.PushEvent:
+		pushEvent = event
+	case *codehub.MergeEvent:
+		mergeEvent = event
+	}
+
+	//触发工作流webhook和测试管理webhook
+	if pushEvent != nil {
+		//add webhook user
+		if len(pushEvent.Commits) > 0 {
+			webhookUser := &commonmodels.WebHookUser{
+				Domain:    req.Header.Get("X-Forwarded-Host"),
+				UserName:  pushEvent.Commits[0].Author.Name,
+				Email:     pushEvent.Commits[0].Author.Email,
+				Source:    setting.SourceFromCodeHub,
+				CreatedAt: time.Now().Unix(),
+			}
+			commonrepo.NewWebHookUserColl().Upsert(webhookUser)
+		}
+
+		//产品工作流webhook
+		if err = TriggerWorkflowByCodehubEvent(pushEvent, baseURI, requestID, log); err != nil {
+			errorList = multierror.Append(errorList, err)
+		}
+
+		if err = updateServiceTemplateByCodehubPushEvent(pushEvent, log); err != nil {
+			errorList = multierror.Append(errorList, err)
+		}
+	}
+
+	if mergeEvent != nil {
+		//产品工作流webhook
+		if err = TriggerWorkflowByCodehubEvent(mergeEvent, baseURI, requestID, log); err != nil {
```

Comment:
improve this if since Trigger... will be called no matter it is a push or a merge